apm-agent-plugins/apm-api-plugin/src/main/java/co/elastic/apm/agent/plugin/api/HeadersExtractorBridge.java

```diff
@@ -34,65 +34,83 @@
 import java.util.Iterator;
 
 @VisibleForAdvice
-public class HeadersExtractorBridge implements TextHeaderGetter<Object> {
+public class HeadersExtractorBridge implements TextHeaderGetter<HeadersExtractorBridge.Extractor> {
 
     private static final Logger logger = LoggerFactory.getLogger(HeadersExtractorBridge.class);
+    @Nullable
+    private static HeadersExtractorBridge INSTANCE;
 
-    private static final HeadersExtractorBridge INSTANCE = new HeadersExtractorBridge();
+    public static class Extractor {
 
-    @VisibleForAdvice
-    public static HeadersExtractorBridge instance() {
-        return INSTANCE;
-    }
+        private static final Extractor INSTANCE = new Extractor();
+        @Nullable
+        private Object headerExtractor;
+        private Object headersExtractor;
 
-    @Nullable
-    private MethodHandle getAllHeadersMethod;
+        private Extractor() {}
 
-    private HeadersExtractorBridge() {
+        public static Extractor of(@Nullable Object headerExtractor, Object headersExtractor) {
```

Comment:
This does not work, unfortunately. There can be multiple implementations passed to the API. Even for the same technology, the `headerExtractor` will be specific to a particular instance of a request object.

Reply:
Of course 🤦‍♂ 
I can make it `ThreadLocal`, or cache per extractor. Since it's our API implementation, I want to make sure it is in accordance to the rest of the stuff we do

Reply:
let's just allocate in this case. There will be allocations anyway for a capturing lambda (like `request::getHeaders`) and the header iterator. Allocations are fine and inevitable in the public API.